numpyro/infer/mcmc.py

```diff
@@ -382,11 +383,74 @@ def _compile(self, rng_key, *args, extra_fields=(), init_params=None, **kwargs):
         except TypeError:
             pass
 
+    @property
+    def init_state(self):
+        """
+        The initial state of the MCMC chain. If this attribute is None,
+        :meth:`run` will call `self.sampler.init(...)` method for initialization.
```

Comment:
My concern with exposing this is that it might give the impression that users can use this to specify initial value of parameters or mass matrix (which isn't true). However, there are other auxiliary variables like `z_grad` and those mentioned in the warning that are non-obvious and would result in leakage of internal details. Could you elaborate on the use case for this?

Reply:
I agree that this warning is tricky to deliver to users. The motivation here is to start the chain on new data with the previously adapted mass matrix, step_size, and the last sample, as requested by @martinjankowiak  and in https://github.com/pyro-ppl/numpyro/issues/534.

For those who want such a warm start feature, they can still achieve that by modifying the warmup state by `state._replace(i=0, pe=1e6)` and taking care of other related diagnostics information (wrong `z_grad` is fine because `z_grad` on the old data only drives momentum on the first leapfrog step - but a wrong pe might lead to rejecting all proposals). So I think we should not expose it.

Reply:
> The motivation here is to start the chain on new data with the previously adapted mass matrix, step_size, and the last sample

Do you mean starting from these values and running adaptation from there? IIUC there are three things that we can specify for initialization - initial values (which is possible using `init_to_value`?), step size and mass matrix. This method is to be able to expose the latter two. Is that correct?

Reply:
Yes, it will re-learn them but started from the previous values. This might be helpful for tricky posteriors, which require spending much time on initial steps...

Reply:
Hmm..how about things like the adaptation window index, wouldn't that need to be reset? 

Another choice might be to have something like `MCMCKernel.get_init_state(self, *args, **kwargs)` method. For HMC, it could be `get_init_state(self, init_params, mass_matrix=None, step_size=None)` which can internally do all this book-keeping (e.g. setting i to 0). That might be a bit more verbose but easier to explain since we can tell the users to use the `get_init_state` method of the corresponding kernel without other caveats.

```python
kernel = HMC(...)
mcmc = MCMC(kernel, ...)
mcmc.initial_state = kernel.get_init_state(init_params, mass_matrix)
mcmc.run(...)
```

Reply:
Makes sense to me. We also need args, kwargs so it is more like `kernel.init` method (except that we won't try to find valid initial params). We have discussed how to expose the inverse mass matrix to the API in https://github.com/pyro-ppl/numpyro/pull/536 but found that it is a bit ambiguous. I think we should skip this feature unless there are more requests for it. What do you think?

> how about things like the adaptation window index

We use [hmc_state.i](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/hmc.py#L314) for the update so it should be fine.

Reply:
I agree, I think it is fine to discuss a bit more before finalizing this. Power users can already do this if they need to, but we should provide an intuitive interface if we are going to advertise this.